lib/reversetunnel/agentpool.go

```diff
@@ -356,12 +366,29 @@ func (l ServerHandlerToListener) Close() error {
 }
 
 func (l ServerHandlerToListener) Addr() net.Addr {
-	return reverseTunnelAddr{}
+	return reverseTunnelAddr(l.tunnelAddr)
+}
+
+type connCloser struct {
+	net.Conn
+	closeOnce *sync.Once
+	closed    chan struct{}
 }
 
+func newConnCloser(c net.Conn) connCloser {
+	return connCloser{Conn: c, closeOnce: new(sync.Once), closed: make(chan struct{})}
+}
+
+func (c connCloser) Close() error {
+	c.closeOnce.Do(func() { close(c.closed) })
```

Comment:
I think this could be simpler:
```go
func (c connCloser) Close() error {
	select {
	case <-c.closed:
  	  	return nil // this way a double Close can be avoided if necessary
	default:
		close(c.closed)
  	  	return c.Conn.Close()
	}
}
```

Reply:
I think your suggestion can still fail when multiple goroutines call `Close` at the same time.
They will both drop into the `default` clause and call `close(c.closed)` causing a panic

Reply:
Ok, that's true.